packages/cli/src/cmds/validator/keymanager/decryptKeystoreDefinitions/index.ts

```diff
@@ -0,0 +1,94 @@
+import {spawn, ModuleThread, Pool, QueuedTask, Worker} from "@chainsafe/threads";
+import {Signer, SignerLocal, SignerType} from "@lodestar/validator";
+import bls from "@chainsafe/bls";
+import {LocalKeystoreDefinition} from "../interface.js";
+import {clearKeystoreCache, loadKeystoreCache, writeKeystoreCache} from "../keystoreCache.js";
+import {lockFilepath, unlockFilepath} from "../../../../util/lockfile.js";
+import {defaultPoolSize} from "./poolSize.js";
+import {DecryptKeystoreWorkerAPI, KeystoreDecryptOptions} from "./types.js";
+
+/**
+ * Decrypt keystore definitions using a threadpool
+ */
+export async function decryptKeystoreDefinitions(
+  keystoreDefinitions: LocalKeystoreDefinition[],
+  opts: KeystoreDecryptOptions
+): Promise<Signer[]> {
+  if (opts.cacheFilePath) {
+    try {
+      const signers = await loadKeystoreCache(opts.cacheFilePath, keystoreDefinitions);
+      if (opts?.onDecrypt) {
+        opts?.onDecrypt(signers.length - 1);
+      }
+      return signers;
+    } catch {
+      // Some error loading the cache, ignore and invalidate cache
+      await clearKeystoreCache(opts.cacheFilePath);
+    }
+  }
+
+  const signers: SignerLocal[] = [];
+  const passwords: string[] = [];
+  const tasks: QueuedTask<ModuleThread<DecryptKeystoreWorkerAPI>, Uint8Array>[] = [];
+  const errors: Error[] = [];
+
+  const pool = Pool(() => spawn<DecryptKeystoreWorkerAPI>(new Worker("./worker.js")), defaultPoolSize);
+  for (const [index, definition] of keystoreDefinitions.entries()) {
+    try {
+      lockFilepath(definition.keystorePath);
+    } catch (e) {
+      if (opts.force) {
+        // Ignore error, maybe log?
+      } else {
+        throw e;
+      }
+    }
+
+    const task = pool.queue((thread) => thread.decryptKeystoreDefinition(definition, Boolean(opts.force)));
+    tasks.push(task);
+    task
+      .then((secretKeyBytes: Uint8Array) => {
+        const signer: SignerLocal = {
+          type: SignerType.Local,
+          secretKey: bls.SecretKey.fromBytes(secretKeyBytes),
+        };
+
+        signers.push(signer);
+        passwords.push(definition.password);
+
+        if (opts?.onDecrypt) {
+          opts?.onDecrypt(index);
+        }
+      })
+      .catch((e: Error) => {
+        // In-progress tasks can't be canceled, so there's a chance that multiple errors may be caught
+        // add to the list of errors
+        errors.push(e);
+        // cancel all pending tasks, no need to continue decrypting after we hit one error
+        for (const task of tasks) {
+          task.cancel();
+        }
+      });
+  }
+
+  try {
+    // only resolves if there are no errored tasks
+    await pool.completed();
```

Comment:
The pool completes when all enqueued tasks resolve or reject. Since the tasks are enqueued synchronously before this, it isn't possible to complete before all tasks are done.